dali/operators/math/expressions/broadcasting.cc

```diff
@@ -274,33 +274,97 @@ void ExpandToNDims(TensorShape<> &sh, int ndim) {
   sh = shape_cat(TensorShape<>(std::vector<int64_t>(ndim - sh.sample_dim(), 1)), sh);
 }
 
-void SimplifyShapesForBroadcasting(TensorShape<>& lhs, TensorShape<> &rhs) {
-  // First, if needed expand dimensions
-  int full_ndim = std::max(lhs.sample_dim(), rhs.sample_dim());
-  if (lhs.sample_dim() != rhs.sample_dim()) {
-    ExpandToNDims(lhs, full_ndim);
-    ExpandToNDims(rhs, full_ndim);
-  }
+void SimplifyShapesForBroadcasting(span<TensorShape<> *> shapes) {
+  int n = shapes.size();
+  SmallVector<TensorShape<>, 3> outs;
+  outs.resize(n);
+
+  int ndim = shapes[0]->size();
```

Comment:
I find it highly confusing that we have `sample_dim` in TensorShape.

Reply:
Still it's dimension of this sample, rather than some arbitrary size. And when you get to TensorShape and TensorListShape residing side by side in the code the usage of `size()` is highly confusing.